File: packages/renderer-android/android/src/main/kotlin/dev/terajs/renderer/android/TerajsHostTransport.kt
package dev.terajs.renderer.android

import org.json.JSONArray
import org.json.JSONObject

object AndroidWireCodec {
  fun decodeCommandBatch(payload: String): List<AndroidHostCommand> {
    val commands = JSONArray(payload)
    return (0 until commands.length()).map { index ->
      val entry = commands.opt(index)
      val objectValue = entry as? JSONObject
        ?: throw IllegalArgumentException("commands[$index] must be an object")
      objectValue.toAndroidHostCommand("commands[$index]")
    }
  }

  fun encodeNativeEventPacket(packet: AndroidNativeEventPacket): String {
    return packet.toJsonObject().toString()
  }
}

class AndroidHostTransport(
  private val applyCommands: (List<AndroidHostCommand>) -> Unit,
  private val emitEventPayload: (String) -> Unit,
  private val diagnostics: AndroidDiagnosticsSink = AndroidLogcatDiagnosticsSink,
) {
  fun receiveCommandBatchPayload(payload: String) {
    val commands = AndroidWireCodec.decodeCommandBatch(payload)
    diagnostics.record(
      AndroidDiagnosticEvent(
        area = "bridge",
        message = "Received command batch",
        details = mapOf("commandCount" to commands.size),
      )
    )
    applyCommands(commands)
  }

  fun makeNativeEventPacketPayload(
    nodeId: Int,
    name: String,
    payload: TerajsJsonValue? = null,
  ): String {
    return AndroidWireCodec.encodeNativeEventPacket(
      AndroidNativeEventPacket(nodeId = nodeId, name = name, payload = payload)
    )
  }

  fun sendNativeEvent(
    nodeId: Int,
    name: String,
    payload: TerajsJsonValue? = null,
  ) {
    diagnostics.record(
      AndroidDiagnosticEvent(
        area = "bridge",
        message = "Sent native event",
        details = mapOf(
          "nodeId" to nodeId,
          "name" to name,
        ),
      )
    )
    emitEventPayload(makeNativeEventPacketPayload(nodeId = nodeId, name = name, payload = payload))
  }
}

File: packages/renderer-ios/ios/Sources/TerajsRendererHost/TerajsHostTransport.swift
import Foundation

public enum TerajsHostTransportError: Error {
  case invalidPayload(String)
  case invalidEncoding
}

public enum UIKitWireCodec {
  public static func decodeCommandBatch(_ payload: String) throws -> [UIKitHostCommand] {
    let data = Data(payload.utf8)
    let json = try JSONSerialization.jsonObject(with: data)
    guard let commands = json as? [Any] else {
      throw TerajsHostTransportError.invalidPayload("Command payload must be a JSON array")
    }

    return try commands.enumerated().map { index, entry in
      guard let object = entry as? [String: Any] else {
        throw TerajsHostTransportError.invalidPayload("commands[\(index)] must be an object")
      }
      return try UIKitHostCommand.fromJSONObject(object, path: "commands[\(index)]")
    }
  }

  public static func encodeNativeEventPacket(_ packet: UIKitNativeEventPacket) throws -> String {
    let data = try JSONSerialization.data(withJSONObject: packet.foundationObject())
    guard let payload = String(data: data, encoding: .utf8) else {
      throw TerajsHostTransportError.invalidEncoding
    }
    return payload
  }
}

public final class UIKitHostTransport {
  public typealias CommandSink = ([UIKitHostCommand]) throws -> Void
  public typealias EventSink = (String) -> Void

  private let applyCommands: CommandSink
  private let emitEventPayload: EventSink

  public init(applyCommands: @escaping CommandSink, emitEventPayload: @escaping EventSink) {
    self.applyCommands = applyCommands
    self.emitEventPayload = emitEventPayload
  }

  public func receiveCommandBatchPayload(_ payload: String) throws {
    try applyCommands(UIKitWireCodec.decodeCommandBatch(payload))
  }

  public func makeNativeEventPacketPayload(
    nodeId: Int,
    name: String,
    payload: TerajsJSONValue? = nil
  ) throws -> String {
    try UIKitWireCodec.encodeNativeEventPacket(
      UIKitNativeEventPacket(nodeId: nodeId, name: name, payload: payload)
    )
  }

  public func sendNativeEvent(
    nodeId: Int,
    name: String,
    payload: TerajsJSONValue? = nil
  ) throws {
    emitEventPayload(try makeNativeEventPacketPayload(nodeId: nodeId, name: name, payload: payload))
  }
}